### PR TITLE
Flag connection as unhealthy when underlying socket is closed.

### PR DESCRIPTION
**Pull request by dave-r12**
https://github.com/square/okhttp/issues/5819

### CODE REVIEW

**Comment by dave-r12** [comment]
I believe this reproduces the issue reported in #5819. This is very subtle. I found that if we close the `rawSocket` in `RealConnection` (using cancel) it is not seen by the application layer `socket` if we are using SSL/TLS. So the health checks pass. When we later call `newCodec` it throws a `SocketException`. Since the connection is considered healthy this continues indefinitely.

File: okhttp/src/main/java/okhttp3/internal/connection/RealConnection.kt [hunk 1]
```diff
@@ -631,9 +631,11 @@ class RealConnection(
   fun isHealthy(doExtensiveChecks: Boolean): Boolean {
     val nowNs = System.nanoTime()
 
+    val rawSocket = this.rawSocket!!
     val socket = this.socket!!
     val source = this.source!!
-    if (socket.isClosed || socket.isInputShutdown || socket.isOutputShutdown) {
+    if (rawSocket.isClosed || socket.isClosed || socket.isInputShutdown ||
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/connection/RealConnection.kt:637]: nice

**Comment by yschimke** [comment]
Impressive detective work.

**Comment by yschimke** [comment]
I wonder whether this could solve a whole bunch of problems after network changes etc.  I'm glad this made 4.5!

**Pull request merged**